### PR TITLE
implemented StateManager

### PR DESCRIPTION
**Pull request by passionfruitjello**

### CODE REVIEW

**Review by bw55555 — APPROVED** [review]
StateManager looks good
Open issue for docstring and tests

File: src/main/java/managers/StateManager.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    public void preInput() {
+        boolean currPreInput = currState.preInput();
+        if (currPreInput) {
+            this.currState = this.nextState("");
+            if (this.currState == null) {
+                this.isDone = true;
+                return;
+            }
+        }
+        this.isDone = false;
```
Comment (mahirhassan-uoft) [inline src/main/java/managers/StateManager.java:28]: this is done

**Review by mahirhassan-uoft — APPROVED** [review]
this is done

**Review by charlietaoxl — APPROVED** [review]
Squeaky clean architecture!

**Pull request merged**